src/ngx_http_lua_socket_tcp.c

```diff
@@ -2027,6 +2032,38 @@ ngx_http_lua_socket_read_all(void *data, ssize_t bytes)
 }
 
 
+static ngx_int_t
+ngx_http_lua_socket_read_bsd(void *data, ssize_t bytes)
+{
+    ngx_http_lua_socket_tcp_upstream_t      *u = data;
+
+    ngx_buf_t                   *b;
+
+#if (NGX_DEBUG)
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, u->request->connection->log, 0,
+                   "lua tcp socket read bsd");
+#endif
+
+    if (bytes == 0) {
+        u->ft_type |= NGX_HTTP_LUA_SOCKET_FT_CLOSED;
+        return NGX_ERROR;
```

Comment:
I'm not sure about this. Maybe it's better to return the empty Lua string when the reading part of the connection has been closed by the remote? This should not be considered a fatal error. And we need a test case to cover this code path.

Reply:
@agentzh 
If we do so, we could not detect if the remote is closed as the same way with `sock:receive('*l')`.

This case is already covered by this line: https://github.com/spacewander/lua-nginx-module/blob/77c34e1f3aec37acb84bf67661ae9cc22b21c2eb/t/058-tcp-socket.t#L3849